mlir/test/Dialect/Affine/unroll.mlir

```diff
@@ -258,6 +259,89 @@ gpu.module @unroll_full {
   }
 }
 
+// UNROLL-FULL-LABEL: func @thread_partial_execution
+func.func @thread_partial_execution() {
+  %0 = arith.constant 0 :index
+  %1 = arith.constant 2 : index    
+  // UNROLL-FULL: %[[C0:.*]] = arith.constant 0 : index
+  gpu.launch blocks(%bx, %by, %bz) in (%sz_bx = %1, %sz_by = %1, %sz_bz = %1)
+             threads(%tx, %ty, %tz) in (%sz_tx = %1, %sz_ty = %1, %sz_tz = %1) {
+    affine.for %iv = %tx to 3 step 2 iter_args(%arg = %0) -> index {
+      %3 = arith.addi %arg, %0 : index
+      affine.yield %3 : index
+    }
+    // UNROLL-FULL: %{{.*}} = affine.for %{{.*}} = %{{.*}} to 3 step 2 iter_args(%[[ARG:.*]] = %[[C0]]) -> (index) {
+    // UNROLL-FULL:   %[[SUM:.*]] = arith.addi %[[ARG]], %[[C0]] : index
+    // UNROLL-FULL:   affine.yield %[[SUM]] : index
+    // UNROLL-FULL: }
+    gpu.terminator
+  }
+  return
+}
+
+// UNROLL-FULL-LABEL: func @invalid_loop
+func.func @invalid_loop() {
+  %0 = arith.constant 0 :index
+  %1 = arith.constant 2 : index
+  gpu.launch blocks(%bx, %by, %bz) in (%sz_bx = %1, %sz_by = %1, %sz_bz = %1)
+             threads(%tx, %ty, %tz) in (%sz_tx = %1, %sz_ty = %1, %sz_tz = %1) {
+    %threadid = gpu.thread_id x
+    affine.for %iv = %tx to 0 step 2 iter_args(%arg = %0) -> index {
```

Comment:
Nit: you forgot to use `%threadid` here

Reply:
The %tx here is actually an abbreviation for thread_id x.

Reply:
* The special case for unroll is that not all threads execute forOp.
```
  std::optional<uint64_t> tripCount = getConstantTripCount(forOp);
  std::optional<uint64_t> maxTripCount = getMaxConstantTripCount(forOp);
```
tripCount = 0
maxTripCount = 1
Keep the loop in this case.

* And the other is the case of dumping the IR in the loop out of the loop.
tripCount = 1
maxTripCount = 1

* core idea
 tripCount = (upper - `(blockSize - 1)`) div stride
 maxTripCount = `(uppper - 0)` div stride
 `(blockSize - 1 )` = maxThreadId = blockSIze - 1
 `0` = minThreadId = 0 
The above rules apply to all scoped Values.

Reply:
It feels like there are no special cases, and all cases are handled by the above formula.
I won't refer to IREE because my project will only rely on LLVM, but ValueBoundsOpInterface might be something to look into, I haven't used it, and by the way, one should think about it from the perspective of threadId, and why gpu.launch should be used here, purely and simply because it can fetch the threadId information.

Reply:
Thank you for your reply, I was about to go to bed, but I'm happy to see your reply and I'm happy to be able to communicate here.

Reply:
What I'm saying is that, instead of trying to update the unroller, you should try to copy this pass from IREE upstream. (Now that I've refactored it to use ValueBoundsOpInterfaec, it could legitimately live upstream, it's just that there's not been any call for that)

Reply:
Do you mean adding a separate pass for removing invalid loops? instead of implementing it here?

Reply:
I thought about it and it would indeed be better to implement this in a separate pass so that people are able to know that there is such a pass as shown in the documentation. thank you for your help for so long (starting with your review of my first PR, and if it is something I understand correctly, I will mention the next PR to implement it.

Reply:
I seem to have misinterpreted what you meant, I thought it would remove invalid loops. It seems to dump the IR in the loop. I think it still makes sense to add a pass like that though.

Reply:
That's the point - if you know a loop will run exactly once, you don't need the loop, right?

Isn't that what you're trying to accomplish here?